cli/internal/run/dry_run.go

```diff
@@ -173,6 +175,43 @@ func executeDryRun(ctx gocontext.Context, engine *core.Engine, g *graph.Complete
 	return taskIDs, nil
 }
 
+func populateCacheState(turboCache cache.Cache, taskSummaries []*taskSummary) error {
+	// We make at most 8 requests at a time for cache state.
+	maxParallelRequests := 8
+	taskCount := len(taskSummaries)
+
+	parallelRequestCount := maxParallelRequests
+	if taskCount < maxParallelRequests {
+		parallelRequestCount = taskCount
+	}
+
+	queue := make(chan int, taskCount)
+
+	var returnErr error
+
+	wg := &sync.WaitGroup{}
+	for i := 0; i < parallelRequestCount; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for index := range queue {
+				task := taskSummaries[index]
+				itemStatus, err := turboCache.Exists(task.Hash)
+				task.CacheState = itemStatus
+				returnErr = err
```

Comment:
Should we be using [`ErrGroup`](https://pkg.go.dev/golang.org/x/sync/errgroup) here? I'm still shakey on what's fine to do in goroutines, but every goroutine writing to the same var seems off.

Reply:
Since I blow up the entire dry run for a single error (this matches current state, BTW) I only need a "has Error" flag. But realizing that clobbering with `nil` is going to end badly.

Reply:
Fixed here: https://github.com/vercel/turbo/pull/3877/commits/42a40c708961c838f5231eda6a68f7c6e20196ac

Good catch.